### ISSUE TITLE
Set number of workers to allocated cgroup CPUs?

### ISSUE THREAD

**Issue by uhthomas** [envoyproxy/envoy#36988]
*Title*: Set number of workers to allocated cgroup CPUs?

*Description*:

I noticed that when running envoy in Kubernetes, it sets the number of workers equal to the number of CPUs on the host, even if there is a limit of just 1 CPU. This may negatively impact performance and it would be nice if envoy could pick this up automatically.

[optional *Relevant Links*:]

https://github.com/golang/go/issues/33803

#5976

**Comment by uhthomas** [comment]
I have tried `--cpuset-threads`, and unfortunately it is still reporting a concurrency of 96. So I guess it will have to be changed manually.

**Comment by uhthomas** [comment]
Is there anyway we can reopen this? Envoy really struggles in environments like this. We observed that upstream connections would overflow and cause disruption. This behaviour is not a great default.

**Comment by uhthomas** [comment]
I hate stale bots :/

**Comment by kyessenov** [comment]
If there is a good way to detect cgroup v2 quota/limit, then it makes a lot of sense to use that.

**Comment by DieBauer** [comment]
I also ran into this issue and worked around this by calculating the 'concurrency' based on cgroupv2. I used https://github.com/uber-go/automaxprocs/ to determine this value. @kyessenov would such an approach (statting the `/sys/fs/cgroup/cpu.max`) seem like 'a good way'? Looking at kernel docs, https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#cpu-interface-files this is the file that does contain the relevant values to determine max concurrency.

**Comment by kyessenov** [comment]
Yes, it would make sense to make Envoy behave rationally on cgroupv2 environments in Kubernetes! Do you know how k8s translates cpu request/limits to the stat files? I am not an expert in this area. If we can reliably read the allocation without some hacks (passing via kubernetes downward APIs) that would be much cleaner. I imagine we could have it by default or via some bootstrap extension, depending on the exact place in the startup sequence when we spawn workers.

**Comment by kyessenov** [comment]
FYI, on Istio, there is one source of truth that will size both envoy and k8s resources consistently via an annotation https://github.com/istio/istio/blob/c48b097b442c721ad526fce8d0e91aa124cf1037/manifests/charts/istio-control/istio-discovery/files/injection-template.yaml#L2. The default concurrency is 2 (which is safe by default).

**Comment by jspaetzel** [comment]
> I have tried `--cpuset-threads`, and unfortunately it is still reporting a concurrency of 96. So I guess it will have to be changed manually.

We had this experience as well... although in our case running 128 by default. Oops. Better defaults here when running in Kubernetes would be great!

**Comment by uhthomas** [comment]
Go 1.25 now respects cgroup cpu limits. Maybe some inspiration? It will make a huge difference for Go deployed in Kubernetes, and this would make a huge difference for envoy in Kubernetes too. 

https://go.dev/doc/go1.25#container-aware-gomaxprocs

**Comment by KBaichoo** [comment]
fyi https://github.com/envoyproxy/envoy/pull/40997 is working on this

**Comment by tsaarni** [comment]
>> Go 1.25 now respects cgroup cpu limits. Maybe some inspiration? 

> fyi [#40997](https://github.com/envoyproxy/envoy/pull/40997) is working on this

Although it was implemented in #40997 and released in version 1.37.0, Envoy does not seem to respect cgroup limits by default, despite what the documentation suggested. I submitted #45410 in hopes to get clarity on this issue.
